File: cpp/src/img_proc/sparse_propagator.cpp
#include <algorithm>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <memory>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <setdetect/img_proc/corners_st.hpp>
#include <setdetect/img_proc/opt_flow_pyr_lk.hpp>
#include <setdetect/img_proc/sparse_propagator.hpp>
#include <setdetect/interpolation/tps_rbf.hpp>
#include <vector>

namespace setdetect::img_proc {

namespace {

auto seed_pts_grid(const cv::Mat& gray, const SparsePropagatorParams& params) -> std::vector<Pt2f> {
    std::vector<Pt2f> pts;
    const int h = gray.rows;
    const int w = gray.cols;
    const int rows = std::max(1, params.seed_rows);
    const int cols = std::max(1, params.seed_cols);
    std::vector<int> row_edges(rows + 1);
    std::vector<int> col_edges(cols + 1);
    for (int ry = 0; ry <= rows; ++ry) {
        row_edges[ry] =
            static_cast<int>(std::round(static_cast<float>(ry) * static_cast<float>(h) / static_cast<float>(rows)));
    }
    for (int cx = 0; cx <= cols; ++cx) {
        col_edges[cx] =
            static_cast<int>(std::round(static_cast<float>(cx) * static_cast<float>(w) / static_cast<float>(cols)));
    }
    for (int ry = 0; ry < rows; ++ry) {
        const int y0 = row_edges[ry];
        const int y1 = row_edges[ry + 1];
        for (int cx = 0; cx < cols; ++cx) {
            const int x0 = col_edges[cx];
            const int x1 = col_edges[cx + 1];
            const int cell_h = y1 - y0;
            const int cell_w = x1 - x0;
            if (cell_h <= 0 || cell_w <= 0) {
                continue;
            }
            const cv::Mat cell = gray(cv::Rect(x0, y0, cell_w, cell_h));
            const std::vector<Corner> corners =
                detect_corners_st(cell, nullptr, params.seed_max_corners_per_cell, params.seed_quality_level,
                                  params.seed_min_distance, 3);
            for (const Corner& c : corners) {
                pts.push_back({static_cast<float>(c.x + x0), static_cast<float>(c.y + y0)});
            }
        }
    }
    return pts;
}

struct MovementResult {
    std::vector<Pt2f> pts;
    std::vector<uint8_t> pts_mask;
    std::unique_ptr<interpolation::ThinPlateSplineRBF> interpolator;

    explicit operator bool() const { return interpolator != nullptr; }
};

auto estimate_movement(const cv::Mat& prev_img, const std::vector<Pt2f>& prev_pts, const cv::Mat& img,
                       const SparsePropagatorParams& params, OpticalFlowCalculator& flow_calc) -> MovementResult {
    MovementResult result;
    if (prev_pts.empty()) {
        return result;
    }

    const std::vector<Pt2f>& prev_pts_col = prev_pts;

    LkOptFlowParams lk_params;
    lk_params.win_size = params.lk_win_size;
    lk_params.max_level = params.lk_max_level;
    lk_params.criteria.max_iter = params.lk_criteria_max_iter;
    lk_params.criteria.eps = params.lk_criteria_eps;

    // Each image's pyramid is built once and reused for both the forward and backward
    // LK passes, instead of calc_optical_flow_pyr_lk rebuilding both images' pyramids
    // on every call (4 builds total for 2 calls, when only 2 are ever needed).
    const OptFlowPyramid prev_pyr = build_opt_flow_pyramid(prev_img, lk_params.max_level, lk_params.win_size);
    const OptFlowPyramid next_pyr = build_opt_flow_pyramid(img, lk_params.max_level, lk_params.win_size);

    std::vector<Pt2f> fwd_pts(prev_pts.size());
    std::vector<uint8_t> fwd_status(prev_pts.size());
    std::vector<float> fwd_err;
    flow_calc.calc_optical_flow(prev_pyr, next_pyr, prev_pts_col, fwd_pts, fwd_status, fwd_err, lk_params);

    std::vector<Pt2f> bwd_pts(fwd_pts.size());
    std::vector<uint8_t> bwd_status(fwd_pts.size());
    std::vector<float> bwd_err;
    // Backward pass: track from img back to prev_img, so the pyramid order is
    // deliberately swapped relative to the forward call above.
    // NOLINTNEXTLINE(readability-suspicious-call-argument)
    flow_calc.calc_optical_flow(next_pyr, prev_pyr, fwd_pts, bwd_pts, bwd_status, bwd_err, lk_params);

    const size_t n = prev_pts.size();
    result.pts.resize(n);
    result.pts_mask.assign(n, 0);

    std::vector<Eigen::VectorXd> rbf_inputs;
    std::vector<Eigen::VectorXd> rbf_outputs;
    rbf_inputs.reserve(n);
    rbf_outputs.reserve(n);

    for (size_t i = 0; i < n; ++i) {
        const float dx = prev_pts[i].x - bwd_pts[i].x;
        const float dy = prev_pts[i].y - bwd_pts[i].y;
        const float fb_error = std::sqrt(dx * dx + dy * dy);
        const bool valid = (fwd_status[i] == 1) && (bwd_status[i] == 1) && (fb_error < params.fb_thresh);
        result.pts[i] = fwd_pts[i];
        result.pts_mask[i] = valid ? 1 : 0;

        if (valid) {
            Eigen::VectorXd in(2);
            Eigen::VectorXd out(2);
            in << fwd_pts[i].x, fwd_pts[i].y;
            out << fwd_pts[i].x - prev_pts[i].x, fwd_pts[i].y - prev_pts[i].y;
            rbf_inputs.push_back(in);
            rbf_outputs.push_back(out);
        }
    }

    if (static_cast<int>(rbf_inputs.size()) >= params.rbf_min_points) {
        if (params.rbf_max_points > 0 && static_cast<int>(rbf_inputs.size()) > params.rbf_max_points) {
            const float step = static_cast<float>(rbf_inputs.size()) / static_cast<float>(params.rbf_max_points);
            std::vector<Eigen::VectorXd> subsampled_in;
            std::vector<Eigen::VectorXd> subsampled_out;
            subsampled_in.reserve(params.rbf_max_points);
            subsampled_out.reserve(params.rbf_max_points);
            for (int k = 0; k < params.rbf_max_points; ++k) {
                const auto idx = static_cast<size_t>(std::floor(static_cast<float>(k) * step));
                if (idx < rbf_inputs.size()) {
                    subsampled_in.push_back(rbf_inputs[idx]);
                    subsampled_out.push_back(rbf_outputs[idx]);
                }
            }
            rbf_inputs.swap(subsampled_in);
            rbf_outputs.swap(subsampled_out);
        }
        try {
            result.interpolator =
                std::make_unique<interpolation::ThinPlateSplineRBF>(rbf_inputs, rbf_outputs, params.rbf_smoothing);
        } catch (...) {
            result.interpolator = nullptr;
        }
    }

    return result;
}

struct RingBuffer {
    std::vector<PropagationStep> buf;
    int size = 0;

    explicit RingBuffer(int capacity) : buf(capacity), size(capacity) {
        for (auto& step : buf) {
            step.frame_id = -1;
        }
    }

    void put(PropagationStep step) {
        const int slot = step.frame_id % size;
        if (slot < 0) {
            return;
        }
        buf[slot] = std::move(step);
    }

    [[nodiscard]] auto get(int frame_id) const -> const PropagationStep* {
        const int slot = frame_id % size;
        if (slot < 0) {
            return nullptr;
        }
        const PropagationStep& step = buf[slot];
        if (step.frame_id != frame_id) {
            return nullptr;
        }
        return &step;
    }
};

// Downscales by scale using INTER_AREA. scale >= 1 is a plain clone: upscaling is
// not a supported use case here, and this keeps the zero-motion first-frame behavior
// exact (no interpolation artifacts from a no-op resize).
auto scale_down(const cv::Mat& src, float scale) -> cv::Mat {
    if (scale >= 1.0F) {
        return src.clone();
    }
    cv::Mat dst;
    cv::resize(src, dst, cv::Size(), scale, scale, cv::INTER_AREA);
    return dst;
}

} // namespace

struct SparsePropagator::Impl {
    SparsePropagatorParams params;
    int frame_id = -1;
    cv::Mat prev_img_scaled;
    std::vector<Pt2f> prev_pts_scaled;
    RingBuffer history;
    OpticalFlowCalculator flow_calc;

    explicit Impl(const SparsePropagatorParams& p) : params(p), history(p.history_size) {}
};

SparsePropagator::SparsePropagator(const SparsePropagatorParams& params) : impl_(std::make_unique<Impl>(params)) {}

SparsePropagator::~SparsePropagator() = default;

SparsePropagator::SparsePropagator(SparsePropagator&&) noexcept = default;
auto SparsePropagator::operator=(SparsePropagator&&) noexcept -> SparsePropagator& = default;

void SparsePropagator::clear() {
    impl_->frame_id = -1;
    impl_->prev_img_scaled = cv::Mat();
    impl_->prev_pts_scaled.clear();
    impl_->history = RingBuffer(impl_->params.history_size);
}

auto SparsePropagator::frame_id() const -> int { return impl_->frame_id; }

auto SparsePropagator::next_img(const cv::Mat& gray) -> PropagationStep {
    cv::Mat img_scaled = scale_down(gray, impl_->params.scale);
    impl_->frame_id += 1;

    const cv::Mat* prev_img_scaled = impl_->prev_img_scaled.empty() ? nullptr : &impl_->prev_img_scaled;
    // Copy previous points before they get overwritten by new seeding
    const std::vector<Pt2f> prev_pts_scaled_copy = impl_->prev_pts_scaled;

    MovementResult movement;
    if (prev_img_scaled != nullptr && !prev_pts_scaled_copy.empty()) {
        movement =
            estimate_movement(*prev_img_scaled, prev_pts_scaled_copy, img_scaled, impl_->params, impl_->flow_calc);
    }

    impl_->prev_img_scaled = std::move(img_scaled);
    impl_->prev_pts_scaled = seed_pts_grid(impl_->prev_img_scaled, impl_->params);

    const float inv_scale = (impl_->params.scale > 0.0F) ? (1.0F / impl_->params.scale) : 1.0F;

    PropagationStep ps;
    ps.frame_id = impl_->frame_id;
    ps.pts.reserve(movement.pts.size());
    ps.prev_pts.reserve(prev_pts_scaled_copy.size());
    for (const Pt2f& p : movement.pts) {
        ps.pts.push_back({p.x * inv_scale, p.y * inv_scale});
    }
    for (const Pt2f& p : prev_pts_scaled_copy) {
        ps.prev_pts.push_back({p.x * inv_scale, p.y * inv_scale});
    }
    ps.pts_mask = movement.pts_mask;
    ps.interpolator = std::shared_ptr<interpolation::ThinPlateSplineRBF>(std::move(movement.interpolator));
    impl_->history.put(ps);
    // Retrieve the stored step to return a valid copy
    const PropagationStep* stored = impl_->history.get(impl_->frame_id);
    if (stored != nullptr) {
        return *stored;
    }
    return PropagationStep{};
}

auto SparsePropagator::propagate(int from_frame_id, std::vector<Pt2f>& pts) const -> bool {
    if (pts.empty()) {
        return true;
    }
    const float scale = impl_->params.scale;
    for (int fid = from_frame_id + 1; fid <= impl_->frame_id; ++fid) {
        const PropagationStep* ps = impl_->history.get(fid);
        if (ps == nullptr || !ps->interpolator) {
            return false;
        }
        try {
            const auto& interpolator = *ps->interpolator;
            for (Pt2f& p : pts) {
                Eigen::VectorXd query(2);
                query << p.x * scale, p.y * scale;
                Eigen::VectorXd disp = interpolator.interpolate(query);
                p.x += static_cast<float>(disp[0]) / scale;
                p.y += static_cast<float>(disp[1]) / scale;
            }
        } catch (...) {
            return false;
        }
    }
    return true;
}

} // namespace setdetect::img_proc

File: py/src/setdetect/card_tracking/sparse_propagator.py
import typing as tp
from dataclasses import dataclass

import cv2
import numpy as np
from numpy.linalg import LinAlgError
from scipy.interpolate import RBFInterpolator


@dataclass
class PropagationStep:
    """Result of one frame consumed by the propagator."""

    frame_id: int  # assigned id for the current frame
    prev_pts: np.ndarray  # points tracked from, in the previous frame
    pts: np.ndarray  # tracked positions in the current frame
    pts_mask: np.ndarray  # which pts are valid
    interpolator: RBFInterpolator | None = None  # movement interpolator


@dataclass
class Movement:
    """Result of optical-flow tracking between two frames."""

    pts: np.ndarray
    pts_mask: np.ndarray
    interpolator: RBFInterpolator | None


def seed_pts(
    img_gray: np.ndarray,
    *,
    rows: int = 8,
    cols: int = 8,
    max_corners_per_cell: int = 8,
    quality_level: float = 0.01,
    min_distance: float = 10.0,
) -> np.ndarray:
    """Find feature points on a uniform grid."""
    pts: list[np.ndarray] = []
    h, w = img_gray.shape[:2]
    row_edges = np.linspace(0, h, rows + 1).astype(int)
    col_edges = np.linspace(0, w, cols + 1).astype(int)
    for ry in range(rows):
        for cx in range(cols):
            cell = img_gray[row_edges[ry] : row_edges[ry + 1], col_edges[cx] : col_edges[cx + 1]]
            corners = cv2.goodFeaturesToTrack(
                cell,
                maxCorners=max_corners_per_cell,
                qualityLevel=quality_level,
                minDistance=min_distance,
            )
            if corners is None:
                continue
            corners = corners.reshape(-1, 2)
            corners[:, 0] += col_edges[cx]
            corners[:, 1] += row_edges[ry]
            pts.append(corners)
    if pts:
        return np.concatenate(pts, axis=0).astype(np.float32)
    return np.empty((0, 2), dtype=np.float32)


def estimate_movement(
    prev_img: np.ndarray,
    prev_pts: np.ndarray,
    img: np.ndarray,
    *,
    lk_win_size: tuple[int, int] = (15, 15),
    lk_max_level: int = 3,
    lk_criteria: tuple[int, int, float] = (cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT, 10, 0.03),
    fb_thresh: float = 1.0,
    rbf_min_points: int = 6,
    rbf_smoothing: float = 1.0,
    rbf_max_points: int = 100,
) -> Movement:
    """Estimate movement between two frames via optical flow."""
    prev_pts_col = prev_pts.reshape(-1, 1, 2)

    # forward flow
    fwd_pts, fwd_status, _ = cv2.calcOpticalFlowPyrLK(
        prev_img,
        img,
        prev_pts_col,
        np.empty_like(prev_pts_col),
        winSize=lk_win_size,
        maxLevel=lk_max_level,
        criteria=lk_criteria,
    )

    # backward flow (forward-backward check)
    bwd_pts, bwd_status, _ = cv2.calcOpticalFlowPyrLK(
        img,
        prev_img,
        fwd_pts,
        np.empty_like(fwd_pts),
        winSize=lk_win_size,
        maxLevel=lk_max_level,
        criteria=lk_criteria,
    )

    # filter
    fb_error = np.linalg.norm(prev_pts - bwd_pts.reshape(-1, 2), axis=1)
    fb_valid = (fwd_status.flatten() == 1) & (bwd_status.flatten() == 1) & (fb_error < fb_thresh)
    pts = fwd_pts.reshape(-1, 2).astype(np.float32)

    # interpolate movement
    displacements = pts - prev_pts
    idx = np.where(fb_valid)[0]
    if rbf_max_points and idx.size > rbf_max_points:
        step = idx.size / rbf_max_points
        idx = idx[np.floor(np.arange(rbf_max_points) * step).astype(int)]
    if idx.size < rbf_min_points:
        interpolator = None
    else:
        try:
            interpolator = RBFInterpolator(
                pts[idx],
                displacements[idx],
                kernel="thin_plate_spline",
                smoothing=rbf_smoothing,
            )
        except LinAlgError:
            interpolator = None

    return Movement(interpolator=interpolator, pts=pts, pts_mask=fb_valid)


class _RingBuffer:
    def __init__(self, size: int) -> None:
        self._size = size
        empty_pts = np.empty((0, 2), dtype=np.float32)
        empty_mask = np.empty(0, dtype=bool)
        self._buf: list[PropagationStep] = [PropagationStep(-1, empty_pts, empty_mask, empty_pts, None)] * size

    @property
    def size(self) -> int:
        return self._size

    def put(self, propagation_step: PropagationStep) -> None:
        slot = propagation_step.frame_id % self._size
        self._buf[slot] = propagation_step

    def get(self, frame_id: int) -> PropagationStep | None:
        slot = frame_id % self._size
        ps = self._buf[slot]
        if ps.frame_id != frame_id:
            return None
        return ps


class SparsePropagator:
    """Establish a frame-to-frame transformation through optical flow."""

    def __init__(
        self,
        *,
        history_size: int = 60,
        scale: float = 0.5,
        seed_kwargs: dict[str, tp.Any] | None = None,
        est_movement_kwargs: dict[str, tp.Any] | None = None,
    ) -> None:
        # parameters
        if seed_kwargs is None:
            seed_kwargs = {}
        if est_movement_kwargs is None:
            est_movement_kwargs = {}
        if history_size < 1:
            raise ValueError(f"buffer_size must be >= 1, got {history_size}")
        if not 0 < scale <= 1:
            raise ValueError(f"scale must be in (0, 1], got {scale}")

        # settings
        self._scale = scale
        self._seed_kwargs = seed_kwargs
        self._est_movement_kwargs = est_movement_kwargs

        # state
        self._frame_id = -1
        self._prev_img_scaled: np.ndarray | None = None
        self._prev_pts_scaled = np.empty((0, 2), dtype=np.float32)
        self._history = _RingBuffer(history_size)

        # helper
        self._no_movement = Movement(
            interpolator=None, pts=np.empty((0, 2), dtype=np.float32), pts_mask=np.empty(0, dtype=bool)
        )

    @property
    def frame_id(self) -> int:
        return self._frame_id

    def clear(self) -> None:
        self._frame_id = -1
        self._prev_img_scaled = None
        self._prev_pts_scaled = np.empty((0, 2), dtype=np.float32)
        self._history = _RingBuffer(self._history.size)

    def _to_scale(self, img_gray: np.ndarray) -> np.ndarray:
        if self._scale == 1.0:
            return img_gray
        return cv2.resize(img_gray, None, fx=self._scale, fy=self._scale, interpolation=cv2.INTER_AREA)

    def _seed_pts_scaled(self, img_scaled: np.ndarray) -> np.ndarray:
        return seed_pts(img_scaled, **self._seed_kwargs).astype(np.float32)

    def next_img(self, img_gray: np.ndarray) -> PropagationStep:
        img_scaled = self._to_scale(img_gray)
        self._frame_id += 1

        prev_img_scaled = self._prev_img_scaled
        prev_pts_scaled = self._prev_pts_scaled

        if prev_img_scaled is None or len(prev_pts_scaled) == 0:
            movement = self._no_movement
        else:
            movement = estimate_movement(prev_img_scaled, prev_pts_scaled, img_scaled, **self._est_movement_kwargs)

        # reseed points for movement estimation
        self._prev_img_scaled = img_scaled
        self._prev_pts_scaled = self._seed_pts_scaled(img_scaled)

        ps = PropagationStep(
            frame_id=self._frame_id,
            pts=movement.pts / self._scale,
            pts_mask=movement.pts_mask,
            prev_pts=prev_pts_scaled / self._scale,
            interpolator=movement.interpolator,
        )
        self._history.put(ps)
        return ps

    def propagate(self, from_frame_id: int, pts: np.ndarray) -> np.ndarray | None:
        scale = self._scale
        for fid in range(from_frame_id + 1, self._frame_id + 1):
            ps = self._history.get(fid)
            if ps is None or ps.interpolator is None:
                return None
            try:
                pts += ps.interpolator(pts * scale).astype(np.float32) / scale
            except LinAlgError:
                return None
        return pts
